### PR TITLE
KD-4141-category-timeout. Requires syspref BorrowerCategoryTimeout (y…

### PR DESCRIPTION
**Pull request by juhak18**
Requires new syspref BorrowerCategoryTimeout. Value in database must be in yaml format. Value should include timeout for category DEFAULT. 
Example:

---
DEFAULT: 1800
USERCATEGORY1: 600
AUTOMAT: 31536000

This case user's in category USERCATEGORY1 have 10 minutes timeout. 
If syspref BorrowerCategoryTimeout does not exist, timeout gets value of syspref timeout.

### CODE REVIEW

**Comment by johannaraisa** [comment]
Is this part working propelly: $timeout =  C4::AuthExtra::get_timeout(undef,$timeout);?
Should the config variable be named differently?

File: C4/AuthExtra.pm [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+    if($error) {
+        $retval = $timeout;
+    }
+
+    return($retval);
+}
+
+
+##############################################
```
Comment (kohaputti) [inline C4/AuthExtra.pm]: What's the purpose of these # characters here?

File: C4/AuthExtra.pm [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    my $error;
+    my ($cardnumber,$timeout) = @_;
+    
+    my $dbh   = C4::Context->dbh;
+    my $valueyaml;
+    my $retval = $timeout;
+ 
+    try {
+
+        #get catgorycode
```
Comment (kohaputti) [inline C4/AuthExtra.pm]: Instead of commenting there could be a function called get_categorycode like you have get_timeout.

File: C4/Auth.pm [hunk 3]
```diff
@@ -1687,6 +1700,11 @@ sub check_cookie_auth {
         my $ip       = $session->param('ip');
         my $lasttime = $session->param('lasttime');
         my $userid   = $session->param('id');
+
+        ####################################
```
Comment (kohaputti) [inline C4/Auth.pm]: Why these # characters?

File: C4/Auth.pm [hunk 4]
```diff
@@ -601,6 +603,7 @@ sub get_template_and_user {
         }
     }
 
+
```
Comment (kohaputti) [inline C4/Auth.pm]: Do these changes in another commit to not mix up with the actual changes

**Review by ghost — COMMENTED** [review]
Here are some suggestions. Also the commit title should be "KD-4141: XXXXXXX" so that the atomic updates work because it reads git history and tries to parse the commit title.

File: C4/AuthExtra.pm [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+            my $sth=$dbh->prepare($sql);
+            $sth->execute();        
+            while (@row=$sth->fetchrow_array()) {
+                $categorycode=$row[0];
+            }
+            $sth->finish();
+        } 
+        #get timeout's valueyaml from systempreferences
+        $sql="select value from systempreferences where variable='BorrowerCategoryTimeout'";
+        $sth=$dbh->prepare($sql);
```
Comment (kohaputti) [inline C4/AuthExtra.pm]: Space is missing at many places around =

Comment (johannaraisa) [reply C4/AuthExtra.pm]: This is a bit unnecessary micromanagering. The job will get done with or without spaces.

Comment (kohaputti) [reply C4/AuthExtra.pm]: Then there will be no chance of getting this to the community

File: C4/AuthExtra.pm [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+            my $sql="select categorycode from borrowers where trim(cardnumber)='".$cardnumber."'";
+            my $sth=$dbh->prepare($sql);
+            $sth->execute();        
+            while (@row=$sth->fetchrow_array()) {
+                $categorycode=$row[0];
+            }
+            $sth->finish();
+        } 
+        #get timeout's valueyaml from systempreferences
+        $sql="select value from systempreferences where variable='BorrowerCategoryTimeout'";
```
Comment (kohaputti) [inline C4/AuthExtra.pm]: Koha::Config::SysPrefs->find() should be used instead of raw SQL which can cause errors more likely

**Comment by johannaraisa** [comment]
There probably should be the atomicupdate (installer/data/mysql/atomicupdate/KD-4141-category-timeout.perl) which creates the BorrowerCategoryTimeout config, also adding the same row creation to "installer/data/mysql/syspref.sql". By naming the commit again like Joonas suggested ensures that the update will run. If you would do these few changes then this is ok.

**Comment by juhak18** [comment]
What if that atomicupdate also fills the BorrowerCategoryTimeout? It could add the old timeout value to all existing categories and to default category.

**Comment by kohaputti** [comment]
> What if that atomicupdate also fills the BorrowerCategoryTimeout? It could add the old timeout value to all existing categories and to default category.

That doesn't sound like it is absolutely needed to be done since one can add those manually. I would only add absolutely minimum amount of changes to Koha-suomi Koha.

**Comment by johannaraisa** [comment]
> > What if that atomicupdate also fills the BorrowerCategoryTimeout? It could add the old timeout value to all existing categories and to default category.
> 
> That doesn't sound like it is absolutely needed to be done since one can add those manually. I would only add absolutely minimum amount of changes to Koha-suomi Koha.

Empty preference on database update is enough, if needed then admins can add the values.
